### PR TITLE
add support for S3 and gc urls to load zarr

### PR DESCRIPTION
**Pull request by toloudis**
Review size: small/med

Allow "Open URL" to load urls to Zarr data that start with "s3:" or "gc:".

A big piece of this code change is just moving some string utility functions into their own shared module (without any modifications).

Otherwise the key implementation is in FileReaderZarr.cpp.  It splits up the url and tells `tensorstore` what kind of underlying "kvstore driver" to use.

### CODE REVIEW

File: agave_app/loadDialog.cpp [hunk 1]
```diff
@@ -32,7 +32,7 @@ LoadDialog::LoadDialog(std::string path, const std::vector<MultiscaleDims>& dims
   std::string title;
   QString pathString = QString::fromStdString(path);
   QString extension = pathString.sliced(pathString.lastIndexOf("."));
-  if (pathString.indexOf("http") == 0) {
+  if ((pathString.indexOf("http") == 0) || (pathString.indexOf("s3:") == 0) || (pathString.indexOf("gs:") == 0)) {
```
Comment (ShrimpCryptid) [inline agave_app/loadDialog.cpp:35]: It says `gc:` in the description for the PR, but `gs:` here? Which one is right?

File: renderlib/StringUtil.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  else
+    return false;
+}
+
+bool
+endsWith(std::string const& value, std::string const& ending)
+{
+  if (ending.size() > value.size())
+    return false;
+  return std::equal(ending.rbegin(), ending.rend(), value.rbegin());
```
Comment (ShrimpCryptid) [inline renderlib/StringUtil.cpp:35]: Instead of searching at `value.rbegin()`, should this start at an offset `ending.size` characters from the end of `value`?

Something like
`return std::equal(ending.rbegin(), ending.rend(), value.rend() - (ending.rend() - ending.rbegin()))`

File: renderlib/io/FileReaderZarr.cpp [hunk 3]
```diff
@@ -16,16 +17,62 @@
...
+        path += "/";
+      }
+      path = path + subpath;
+      if (!endsWith(path, "/")) {
+        path += "/";
+      }
+      return { { "driver", "gcs" }, { "bucket", bucket }, { "path", path } };
+    } else {
+      return { { "driver", "http" }, { "base_url", filepath }, { "path", subpath } };
+    }
```
Comment (ShrimpCryptid) [inline renderlib/io/FileReaderZarr.cpp:75]: Looks good! I'd love to see some unit tests for this though

**Review by ShrimpCryptid — COMMENTED** [review]
Couple small comments. I think the string methods would really benefit from unit tests!

File: renderlib/StringUtil.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (toloudis) [reply renderlib/StringUtil.cpp:35]: rbegin and rend are iterators that start from the END of the string. The `r` means reverse.  So rbegin starts at the end and iterates backward to rend (which represents the beginning!)  And in the flavor of std::equal that takes 3 args, the second range is taken to have the same number of elements as the first range.  In this case that means the length of the `ending` string.

File: renderlib/io/FileReader.cpp [hunk 4]
```diff
@@ -42,6 +42,10 @@ FileReader::getReader(const std::string& filepath, bool isImageSequence)
     return new FileReaderImageSequence(filepath);
   } else if (filepath.find("http") == 0) {
     return new FileReaderZarr(filepath);
+  } else if (filepath.find("s3:") == 0) {
+    return new FileReaderZarr(filepath);
+  } else if (filepath.find("gs:") == 0) {
+    return new FileReaderZarr(filepath);
```
[the remark this answers is not in the document]
Comment (toloudis) [reply renderlib/io/FileReader.cpp:48]: Might as well..  in this case the expression is so short, it makes little difference but why not!

**Comment by toloudis** [comment]
> Couple small comments. I think the string methods would really benefit from unit tests!

Good idea, I can throw some in!

File: renderlib/io/FileReader.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (toloudis) [reply renderlib/io/FileReader.cpp:48]: also bear in mind that this StringUtils was already working code in the places where it was used.  I am just moving it into a common file for this pr.

File: test/test_stringUtil.cpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    split("hello world", 'x', elems);
+    REQUIRE(elems.size() == 1);
+    REQUIRE(elems[0] == "hello world");
+
+    elems.clear();
+    split("hello world", 'o', elems);
+    REQUIRE(elems.size() == 3);
+    REQUIRE(elems[0] == "hell");
+    REQUIRE(elems[1] == " w");
+    REQUIRE(elems[2] == "rld");
```
Comment (ShrimpCryptid) [inline test/test_stringUtil.cpp:61]: hell world 😱

File: test/test_stringUtil.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    REQUIRE(elems.size() == 2);
+    REQUIRE(elems[0] == "name");
+    REQUIRE(elems[1] == "value");
+
+    elems.clear();
+    split("name=value\nname2=value2\n", '\n', elems);
+    REQUIRE(elems.size() == 2);
+    REQUIRE(elems[0] == "name=value");
+    REQUIRE(elems[1] == "name2=value2");
+  }
```
Comment (ShrimpCryptid) [inline test/test_stringUtil.cpp:74]: Hmm, this is interesting to me. I think in JS the expected behavior is `["name=value", "name2=value2", ""]` where the ending `\n` would cause an empty string at the end? Any reason to just outright discard it?

Comment (ShrimpCryptid) [reply test/test_stringUtil.cpp:74]: Especially since it looks like you're adding a check for empty strings in `splitToNameValuePairs`...

File: test/test_stringUtil.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    REQUIRE(pairs.size() == 2);
+    REQUIRE(pairs["name"] == "value");
+    REQUIRE(pairs["name2"] == "value2");
+
+    pairs = splitToNameValuePairs("name=value\nname2=value2\nname3=value3\n\n");
+    REQUIRE(pairs.size() == 3);
+    REQUIRE(pairs["name"] == "value");
+    REQUIRE(pairs["name2"] == "value2");
+    REQUIRE(pairs["name3"] == "value3");
+  }
```
Comment (ShrimpCryptid) [inline test/test_stringUtil.cpp:107]: Maybe add one more test where there's multiple newlines between two elements?

**Review by ShrimpCryptid — COMMENTED** [review]
Ooh, these unit tests are great, thank you! I added some last nit comments, will approve once you respond!

File: test/test_stringUtil.cpp (hunk 6 above)
Comment (toloudis) [reply test/test_stringUtil.cpp:74]: I added an extra test that shows empty string.  This is just how my implementation works.  It's using std::getline with the specified delimiter.  When the delimiter is at the end of the string, you don't get any tokens after that.

File: test/test_stringUtil.cpp (hunk 7 above)
Comment (toloudis) [reply test/test_stringUtil.cpp:119]: ok done!

File: agave_app/loadDialog.cpp (hunk 1 above)
Comment (meganrm) [inline agave_app/loadDialog.cpp:35]: Is there a possibility there will be more of these in the future? Just thinking this looks like a place for a "isURL" function with approved prefixes

Comment (toloudis) [reply agave_app/loadDialog.cpp:35]: I'm not aware of any yet but I agree this is about at the limit where if any new one comes up it will get refactored.

**Review by meganrm — APPROVED** [review]

**Review by ShrimpCryptid — APPROVED** [review]
